src/JustSaying/Messaging/Monitoring/Operation.cs

```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Diagnostics;
+
+namespace JustSaying.Messaging.Monitoring
+{
+    public sealed class Operation : IDisposable
+    {
+        private readonly IMessageMonitor _messageMonitor;
+        private readonly Action<TimeSpan, IMessageMonitor> _onComplete;
+        private readonly Stopwatch _stopWatch;
+
+        internal Operation(IMessageMonitor messageMonitor, Action<TimeSpan, IMessageMonitor> onComplete)
+        {
+            _messageMonitor = messageMonitor ?? throw new ArgumentNullException(nameof(messageMonitor));
+            _onComplete = onComplete ?? throw new ArgumentNullException(nameof(onComplete));
+
+            _stopWatch = new Stopwatch();
+            _stopWatch.Start();
+        }
+
+        public void Dispose()
+        {
+            _onComplete(_stopWatch.Elapsed, _messageMonitor);
```

Comment:
Should the stopwatch be being stopped first? Leaving it running implies it would support `Dispose()` being called multiple times, but then that's a smell of the `IDisposable` pattern as it's not intended to be used that way.

Reply:
I've made `MonitorExtensions` internal as Operation is only used there to provide a nice interface for wrapping timed operations. Making it internal hides the ability to dispose from users, so we don't need to worry.